flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/HsMemoryDataManager.java

```diff
@@ -0,0 +1,673 @@
+            // released buffer is not needed.
+            if (bufferContext.isReleased) {
+                return false;
+            }
+            boolean isNeeded = true;
+            switch (spillStatus) {
+                case NOT_SPILL:
+                    isNeeded = !bufferContext.isSpill;
+                    break;
+                case SPILL:
+                    isNeeded = bufferContext.isSpill;
+                    break;
+            }
+            switch (consumeStatus) {
+                case NOT_CONSUMED:
+                    isNeeded &= !bufferContext.isConsumed;
+                    break;
+                case CONSUMED:
+                    isNeeded &= bufferContext.isConsumed;
+                    break;
+            }
+            return isNeeded;
+        }
+    }
+
+    /**
+     * This class maintains the buffer information and its status for hybrid shuffle mode.
+     *
+     * <p>Note that: This class itself is not thread-safe, guarded by subpartition lock.
+     */
+    private static class BufferContext {
+        private final Buffer buffer;
+
+        private final BufferIndexAndChannel bufferIndexAndChannel;
+
+        private boolean isReleased;
+
+        private boolean isSpill;
+
+        private boolean isConsumed;
+
+        private CompletableFuture<Void> spilledFuture;
+
+        public BufferContext(Buffer buffer, int bufferIndex, int subpartitionId) {
+            this.bufferIndexAndChannel = new BufferIndexAndChannel(bufferIndex, subpartitionId);
+            this.buffer = buffer;
+        }
+
+        public void onBufferReleased() {
+            isReleased = true;
+            buffer.recycleBuffer();
+        }
+
+        public void onBufferSpill(CompletableFuture<Void> spilledFuture) {
+            // retainBuffer and mark buffer is spill status.
+            buffer.retainBuffer();
+            isSpill = true;
+            checkState(this.spilledFuture == null, "Already set spill future for buffer context.");
+            this.spilledFuture = spilledFuture;
+        }
+
+        public void onBufferSpilled() {
+            // decrease ref count when spilling is finished.
+            buffer.recycleBuffer();
+        }
+
+        public void onBufferConsumed() {
+            isConsumed = true;
+            // add ref count when buffer consumed.
+            buffer.retainBuffer();
+        }
+    }
+
+    /** Integrate the buffer and dataType of next buffer. */
+    public static class BufferAndNextDataType {
+        private final Buffer buffer;
+
+        private final Buffer.DataType nextDataType;
+
+        public BufferAndNextDataType(Buffer buffer, Buffer.DataType nextDataType) {
+            this.buffer = buffer;
+            this.nextDataType = nextDataType;
+        }
+
+        public Buffer getBuffer() {
+            return buffer;
+        }
+
+        public Buffer.DataType getNextDataType() {
```

Comment:
```suggestion
        public Optional<Buffer.DataType> getNextDataType() {
```

Reply:
I think `nextDataType` should never be null, if a buffer without next buffer, nextDataType of this buffer will be `Buffer.DataType.NONE`.